pkg/namespace/namespace.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package namespace
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"syscall"
+
+	"github.com/prometheus/procfs"
+
+	"github.com/parca-dev/parca-agent/pkg/cgroup"
+)
+
+// PIDNamespaceAdjacentPIDs returns the PIDs of processes that share the same PID namespace and the same cgroup.
+func PIDNamespaceAdjacentPIDs(pid int) ([]int, error) {
+	proc, err := procfs.NewProc(pid)
+	if err != nil {
+		return nil, err
+	}
+	namespaces, err := proc.Namespaces()
+	if err != nil {
+		return nil, err
+	}
+	inodes := map[uint32]struct{}{}
+	for _, namespace := range namespaces {
+		if namespace.Type == "pid" || namespace.Type == "pid_for_children" {
+			inodes[namespace.Inode] = struct{}{}
+		}
+	}
+	nsInodes := []uint32{}
+	for inode := range inodes {
+		nsInodes = append(nsInodes, inode)
+	}
+
+	cgs, err := proc.Cgroups()
+	if err != nil {
+		return nil, err
+	}
+	cg := cgroup.FindContainerGroup(cgs)
+
+	procs, err := procfs.AllProcs()
+	if err != nil {
+		return nil, err
+	}
+	adjPIDs := map[int]struct{}{}
+	for _, p := range procs {
+		if p.PID == pid {
+			continue
+		}
+		namespaces, err := p.Namespaces()
+		if err != nil {
+			return nil, err
+		}
+
+		cgs, err := p.Cgroups()
+		if err != nil {
+			return nil, err
+		}
+		ccg := cgroup.FindContainerGroup(cgs)
+
+		for _, namespace := range namespaces {
+			for _, inode := range nsInodes {
+				if namespace.Inode == inode && ccg.Path == cg.Path {
```

Comment:
Nicely done! This should handle the case where containers share the same PID namespace

https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/